mono/metadata/metadata.c

```diff
@@ -3019,25 +3019,30 @@ mono_metadata_get_generic_inst (int type_argc, MonoType **type_argv)
 
 	collect_data_free (&data);
 
-	mono_image_set_lock (set);
+	ginst = (MonoGenericInst *)mono_conc_hashtable_lookup (set->ginst_cache, ginst);
+	if (ginst)
+		return ginst;
 
-	ginst = (MonoGenericInst *)g_hash_table_lookup (set->ginst_cache, ginst);
-	if (!ginst) {
-		ginst = (MonoGenericInst *)mono_image_set_alloc0 (set, size);
+	ginst = (MonoGenericInst *)mono_image_set_alloc0 (set, size);
 #ifndef MONO_SMALL_CONFIG
-		ginst->id = ++next_generic_inst_id;
+	ginst->id = InterlockedIncrement (&next_generic_inst_id);
 #endif
-		ginst->is_open = is_open;
-		ginst->type_argc = type_argc;
-
-		for (i = 0; i < type_argc; ++i)
-			ginst->type_argv [i] = mono_metadata_type_dup (NULL, type_argv [i]);
+	ginst->is_open = is_open;
+	ginst->type_argc = type_argc;
 
-		g_hash_table_insert (set->ginst_cache, ginst, ginst);
-	}
+	for (i = 0; i < type_argc; ++i)
+		ginst->type_argv [i] = mono_metadata_type_dup (NULL, type_argv [i]);
 
+	mono_image_set_lock (set);
+	MonoGenericInst *ginst2 = (MonoGenericInst *)mono_conc_hashtable_insert (set->ginst_cache, ginst, ginst);
 	mono_image_set_unlock (set);
-	return ginst;
+
+	if (!ginst2)
+		 ginst2 = ginst;
+	else
+		free_generic_inst (ginst);
```

Comment:
Can this be added as an api comment ? Normal hashtable inserts replace the old values.